### ISSUE TITLE
Failed to parse transaction with output with OP_RETURN

### ISSUE THREAD

**Issue by humbcezar** [bitcoin-education/bitcoin-java#22]
When calling Transaction.fromByteStream with the content below (in hex), it raises  java.lang.RuntimeException: Parsing script failed error. Testnet tx id: b4c38763bdb2a310664dac732e899bbf40389ed52bb1100a7ee5cc06a7e534b8

020000000001010000000000000000000000000000000000000000000000000000000000000000ffffffff4803b00921046aea2462425443506f6f6cfabe6d6dbb850e610863e6aed345a8681b41919e7df23f04cab5cc8cd3d392bd917e959a040000000d6531d503040db1640d010000000000ffffffff02c69e4b000000000017a9147bef0b4a4dafa77b2ec52b81659cbcf0d9a91487870000000000000000266a24aa21a9ed90bb8a0549f0e2594d8cf989cc0d53acef0f281ba6e7b194fcf343c7f0663b1a0120000000000000000000000000000000000000000000000000000000000000000000000000

**Comment by humbcezar** [comment]
Other tx hex examples:

020000000001010000000000000000000000000000000000000000000000000000000000000000ffffffff4803b1092104c0ec2462425443506f6f6cfabe6d6d3522eded34ea9aa46a6c220266b109ccb25840e5befbc0489bbd24105e6c1ccf040000000d6531d503040db11203010000000000ffffffff02df244b000000000017a9147bef0b4a4dafa77b2ec52b81659cbcf0d9a91487870000000000000000266a24aa21a9edbeeb475bd2efa08c5de870bd641137054ad5c78bfe30237995fd74bbac504bec0120000000000000000000000000000000000000000000000000000000000000000000000000

020000000001010000000000000000000000000000000000000000000000000000000000000000ffffffff4803b2092104f5ec2462425443506f6f6cfabe6d6d71b3850167566440b52652625feb1a9ebecce4551c36edff5bd6739ec496471b040000000d6531d503040db1a816000000000000ffffffff0249c94a000000000017a9147bef0b4a4dafa77b2ec52b81659cbcf0d9a91487870000000000000000266a24aa21a9ed1abb071efbbd947b4352daabb91cd3d44a1fafac1b465b49a238217cf09f9fe70120000000000000000000000000000000000000000000000000000000000000000000000000

020000000001010000000000000000000000000000000000000000000000000000000000000000ffffffff4803b3092104b8ee2462425443506f6f6cfabe6d6d62b6e8987ae12a7d5691dbd8f32af7c132c7b4d6b100d0f7b762aa89f7f3f62f040000000d6531d503040db150d1000000000000ffffffff02ea524b000000000017a9147bef0b4a4dafa77b2ec52b81659cbcf0d9a91487870000000000000000266a24aa21a9ed1e1efdbda191a7de9d1f187a2c6ff9c40a47717857f5217de0ed41d8d6bdd8db0120000000000000000000000000000000000000000000000000000000000000000000000000

**Comment by bash83** [comment]
So seems to happen with all coinbase transactions. The coinbase signature is pretty much useless (first few bytes include the block height but everything after it is random text added by miner).

In the example above, the coinbase is:
`03b00921046aea2462425443506f6f6cfabe6d6dbb850e610863e6aed345a8681b41919e7df23f04cab5cc8cd3d392bd917e959a040000000d6531d503040db1640d010000000000`

Where:
`03` is the varint for block size (3)
`b00921` is LE of the block height (2165168)
`046aea2.....0000` is miner text (without size)

The current "Signature" object does not properly handle this. 

See here for more info [https://github.com/bitcoin/bips/blob/master/bip-0034.mediawiki](https://github.com/bitcoin/bips/blob/master/bip-0034.mediawiki)

**Comment by humbcezar** [comment]
> So seems to happen with all coinbase transactions. The coinbase signature is pretty much useless (first few bytes include the block height but everything after it is random text added by miner).
> 
> In the example above, the coinbase is: `03b00921046aea2462425443506f6f6cfabe6d6dbb850e610863e6aed345a8681b41919e7df23f04cab5cc8cd3d392bd917e959a040000000d6531d503040db1640d010000000000`
> 
> Where: `03` is the varint for block size (3) `b00921` is LE of the block height (2165168) `046aea2.....0000` is miner text (without size)
> 
> The current "Signature" object does not properly handle this.
> 
> See here for more info https://github.com/bitcoin/bips/blob/master/bip-0034.mediawiki

Yes. When the transaction input is coinbase (txID = 32 bytes of 0, previous index = 0xffffffff), we have to parse the scriptSig differently.

In order to solve the issue in a decent way, we have to separate the scriptSig and the scriptPubKey in two classes, both inheriting from the Script class. The Script class is already too bloated, there are things in it that only apply to scriptPubKey, so the refactoring can solve this problem too.

I have other priorities now, so I'll only have time to solve this issue in a month or two. In the meantime, the issue stays open if someone else wants to work on it.
